### ISSUE TITLE
ensure versioning/publishing validated properly

### ISSUE THREAD

**Issue by boneskull** [mochajs/mocha#2999]
given the root `mocha.js` dist file shouldn't change until a version bump, and we build a new version of this dist file for tests, we are missing the potential problem where everything is fine for releasing *except* someone forgot to rebuild `mocha.js`.

granted, the `preversion` script should avoid this problem, but there's nothing stopping a maintainer from bumping via other means.

**Comment by ScottFreeCode** [comment]
To be exact, under our current setup it will be fine if:

1. the version is bumped immediately before publishing and
2. the version bump is accomplished via NPM

What we need, at the most fundamental level, is not necessarily an assertion so much as simply something (assertion, or just doing the build) to get this right **in the publishing step**.

**Comment by ScottFreeCode** [comment]
Since version 4, we now build in the `prepublishOnly` step, but only test in the version step, which opens us to more or less the same dangers (changes between version and publish, manually versioning); the remaining thing to do along the lines of this issue would be to move the test from `preversion` to `prepublishOnly`, right?

**Comment by boneskull** [comment]
I'm a bit confused--`prepublishOnly` will fix the original issue here, which was that building the dist file wasn't necessarily happening.

**Comment by ScottFreeCode** [comment]
Yeah -- we could close this, since for using for using an up-to-date built file we've got it solved.

I'm just saying that testing (ensuring that the built file is going to be any good) still has the same basic issue of not necessarily happening because it's associated with NPM-based versioning. If we closed this I'd want to open a near-identical issue with "test" replacing "build".

**Comment by boneskull** [comment]
Right.  My motivation to remove it was that I ran `npm version blah` which ran the tests, pushed, asserted the build passed, then was gonna run `npm publish` which would just run the tests again, knowhatImean?  Just seemed like a waste of time at that point.  Perhaps removing the `preversion` script would be fine, since you don't necessarily have to *push* once you've done that--but `npm publish` makes things happen immediately!

I'll make this change

**Comment by boneskull** [comment]
Oh, also, I noticed `npm version` will actually fail in npm@5.4.2 if `.package-lock.json` is in your `.gitignore`...

**Comment by ScottFreeCode** [comment]
> Oh, also, I noticed  npm version  will actually fail in npm@5.4.2 if  .package-lock.json  is in your  .gitignore ...

Wait, so if we use `package-lock.json` we break CI installs due to faulty dependency trees, but if we don't use `package-lock.json` we *have* to version manually? That's one I hadn't heard of before.

**Comment by boneskull** [comment]
I really dunno.  I'll give `package-lock.json` another shot here soon.  could be a travis-related caching problem.

**Comment by boneskull** [comment]
I'm reopening this as we had another publishing problem that this fix didn't address: missing file(s) in the published package.

As part of our "smoke" tests, we need to ensure the output of (perhaps?) `npm pack` works properly. Open to other ideas, but this is my first thought.  Create a script to:

1. Run `npm start build && npm pack --json`
1. Parse the result of `npm pack --json` to get the output filename
1. Unarchive the file to some temp directory
1. Run `/path/to/temp/bin/mocha --no-config --reporter spec test/sanity/sanity.spec.js`

This script should *replace* the `script` property of the `smoke` stage in `.travis.yml`.  It should be written in JavaScript.

**Comment by boneskull** [comment]
Rename "sanity" to "smoke" while we're at it.

**Comment by mark-wiemer** [comment]
Releases have been automated for a long while now :)
